### ISSUE TITLE
Make redirects more consistent

### ISSUE THREAD

**Issue by guilhermednt** [redelivre/login-cidadao#44]
To avoid redirect loops like the one reported at #37 I propose we implement a more consistent way of handling redirects.

As I see it, we have two types of redirect:

- the "after login" redirect: this happens when the User tries to access the authorization endpoint, for example, but is not authenticated. The originally requested URL is saved and the user is redirected to it after logging in;
- the "task" redirect: imagine the previous scenario where the User is trying to authorize a RP but after logging in we find out the Terms of Service changed and the User has to accept them before using any other feature.

Currently, if there is a need to do a "task" redirect we simply lose the original request, making the user land into an undesired page thus breaking the experience of whatever the person was trying to accomplish.

My proposal consists of implementing an `intent` value that should store the user's ultimate intent for the current session (in the example: accessing the authorization endpoint) and that `intent` would be consumed only after there are no pending `tasks`. Bellow is a very simple flowchart of the proposed.

![selection_788](https://cloud.githubusercontent.com/assets/860796/16988060/87f74b6c-4e65-11e6-8c43-f499c8df5135.png)

It's worth noting that the `task` should have some kind of attribute/metadata to determine it's priority, since the Terms of Service acceptance seems to be more important than the user updating a password. There would be no point in asking someone to update their password to, after that, show that the TOS changed and they do not agree with the changes.

Likewise, we could also implement some kind of task limit per session to avoid overwhelming the user with several repetitive tasks, so tasks that are less important could be postponed to the next opportunity.

**Comment by guilhermednt** [comment]
Currently, IIRC, we have the following `tasks`, in descending order of priority:

1. Session Invalidation
2. TOS Agreement
3. Email confirmation
4. Password Algorithm Migration
